### DISCUSSION TITLE
z80pack 1.38 release

### DISCUSSION THREAD

**Discussion by udo-munk** [udo-munk/z80pack#198] [General]
There have been so many improvements and additions since the 1.37 release, I really would like to create the 1.38 final release from the development branch in the near future. If you still need to get anything into 1.38-dev please discuss here, so that I get an idea when I can start preparing the release, thanks.

**Comment by sneakywumpus** [comment]
I opened PR #199 to clean up load_file().

**Reply by udo-munk** [reply]
I also get an email for an opened PR.

**Comment by sneakywumpus** [comment]
I'm adding a macro table to the list file.

**Comment by sneakywumpus** [comment]
Just some questions:
Was it intentional that the following files use 16 bytes per line?
I use a little script to check that I didn't break the assembler that assembles all .asm files and compares the .hex and .lis output.
Do you use the same -e option for all .asm files?

altairsim/roms/turnmon.hex
imsaisim/life.hex
imsaisim/microchess.hex
imsaisim/scs1.hex

**Reply by udo-munk** [reply]
No it was not intentional, I used 16 bytes per line to compare with the old version and for some I forgot to assemble again with the default. I'll replace these files.

**Comment by sneakywumpus** [comment]
I have locally changed the macro code to not convert to upper case. This looks much nicer in the listing,
but causes a problem for the XYBASIC code. It has a macro "error" with creates character constants from
parameters that are not character constants. This causes a divergence from the code produced by MAC
(which convert everything to upper case). The programmers seem to have thought about this, since in the
70 uses of this macro only 4 calls have lower case parameters. What to do? Change the 4 lines in XYBASIC,
or keep the conversion the upper case in the macro code?

**Reply by sneakywumpus** [reply]
The macro code shouldn't convert to upper case, since this makes it impossible to use lower case with macros.
We could add a flag to the assembler to force conversion to upper case.

**Reply by sneakywumpus** [reply]
The only character case specific thing in the assembler before the introduction of macros was
that symbols are case insensitive and are listed as upper case in the symbol table.
Lets keep it that way. So, introduce a flag for upper conversion or change the affected assembler
source code?

**Reply by udo-munk** [reply]
I would leave the source code as is, because the way it is it can be build on CP/M and ISIS systems. Would be better to have assembler options for the wanted behaviour.

**Reply by sneakywumpus** [reply]
This wouldn't change the possibility to build on CP/M or ISIS systems.
These are the changes that would be needed:
```
diff ./bendix ../bendix
117c117
< 	error	f,d,k		;issue fatal DK error
---
> 	error	f,D,K		;issue fatal DK error
diff ./eval ../eval
620c620
< 	error	n, o, v		;else nonfatal OV error
---
> 	error	n, O, V		;else nonfatal OV error
diff ./fp9511 ../fp9511
228c228
< fover:	error	n, o, v		;issue nonfatal OV error
---
> fover:	error	n, O, V		;issue nonfatal OV error
diff ./packard ../packard
84c84
< 	error	f, f, i		;and issue fatal FI error
---
> 	error	f, F, I		;and issue fatal FI error
```

All other uses of `error` already use upper case for the 2nd and 3rd parameter.
The files `bendix` and `packard` aren't used for the CP/M, Altair and Imsai builds,
I included these just for completeness.

**Reply by udo-munk** [reply]
OK, these modules were customizations for special customer machines used as health monitor, no one would build theses versions anymore, so not really essential.

**Comment by ratboy666** [comment]
I always thought that the macro (and label, etc) lookup were not case
sensitive... but, case
was to be retained in 'quoted' strings. Anywhere else, a conversion to
upper could be
expected. The only problem is in the IRPC macro. So, let us run some
code through M80:

.8080
0000' aseg
org 100h
irpc c,abcABCxyz
db '&c'
endm
0100 61 + db 'a'
0101 62 + db 'b'
0102 63 + db 'c'
0103 41 + db 'A'
0104 42 + db 'B'
0105 43 + db 'C'
0106 78 + db 'x'
0107 79 + db 'y'
0108 7A + db 'z'
end

And here we see that the string is NOT upper-cased, but taken as is.
Now, if

db '&c'

is replaced by

db &c

we see that

db a is the SAME as db A -- which is to say 07 (the value of A as a
register operand).

That compare is done case insensitive -- so, delay the compare until
the actual symbol table lookup, but store
in mixed case (for IRPC expansion). Note... as far as I remember, this
was also the Intel Assembler behaviour, but
it has been a long time... For macros, we should strive to be as
consistent as possible! One case-fold too many...
and there may be a macro that cannot be expressed! But... in this case,
not that A evals to 7, and to force that
down one macro expansion, use %A, which is replaced with 7 -- her is
the example:

mymac macro p
  db '&p'
endm

mymac %A

which results in:

db '7'

because... %A replaces %A with 7, which is then passed to mymac... This
can be a life-saver in macros, because x set x+1
then macrocall %x passes the value of X, not the symbol. Thus allowing
us to implement stacks, and data structures
in macro. But... this also implies that symbol and SYMBOL should be the
same. Except... for IRPC (as seen above). Note
also that '&param' in the macro needs to recover the case of each of
the characters.

The macro code itself should, in general, not convert to upper case --
can you give me that actual problematic
macro for XYBASIC?

Thanks,
Fred Weigel

**Comment by sneakywumpus** [comment]
```
error   macro   typ, m1, m2
        call    erro&TYP
        db      '&M1', '&M2' or 80H
        endm
```
Normally this would be called with, for example, `error n, O, V` but in a few places they put `error n, o, v`,
so you get lower case character constants when the assembler is not converting to upper case like DRI MAC.

**Comment by sneakywumpus** [comment]
I think the assembler is in a release candidate state now.
If anyone else is listening: please test the assembler from the "dev" branch,
especially the new macro functions and expression parser.

**Reply by sneakywumpus** [reply]
Now I'm really running out of ideas of what can be done to the assembler.
Only testing remains.

**Comment by sneakywumpus** [comment]
I was thinking about moving the common stuff from sim.h into z80core as simcore.h and replace it with an ```#include "simcore.h"``` at the end of each sim.h.

**Reply by udo-munk** [reply]
As long as you don't move anything in the user definable section that is ok. There should be only one file where the user customizes a machine to their needs.

**Reply by sneakywumpus** [reply]
Should BUS_8080 be configurable or solely depend on FRONTPANEL?

**Comment by udo-munk** [comment]
Doesn't need to be configurable, the CPU bus signals are provided for FRONTPANEL only.

**Comment by sneakywumpus** [comment]
Is there a particular reason why the Makefile's specify gcc and g++ as the compiler?
Wouldn't be better to just leave CC and CXX at their system default values, as the civetweb Makefile does?
So that one doesn't need, for example, have to explicitly type "make CC=cc" in cpmsim/srctools on FreeBSD.

**Reply by sneakywumpus** [reply]
I mean nowadays on every *nix like system "cc" is either gcc or clang.
I think the gcc usage comes from the 80s/90s when vendor compilers didn't cut it or cost extra and
nearly everyone installed gcc on their systems.

**Reply by udo-munk** [reply]
Yes, might be better to leave CC and CCX at the system defaults, one can override it with make CC=whatever anyway.

**Comment by sneakywumpus** [comment]
While building z80pack on FreeBSD I discovered that there is a package for z80pack. This package is in
conflict with another package that includes hex2bin and bin2hex utilities, so I'm thinking about renaming
bin2hex to cpmb2h and ptp2bin to cpmp2b to avoid any possible conflicts.

**Reply by udo-munk** [reply]
Not really great names, hasn't to do with CP/M, these are generic format conversion programs.

**Reply by sneakywumpus** [reply]
Do you have any suggestions, or just leave it as it is?

**Reply by udo-munk** [reply]
Just leave as is.

**Comment by sneakywumpus** [comment]
I'm done with the assembler, and the simulators now compile cleanly and run on FreeBSD,
so I won't hinder your release process anymore :-)

**Reply by udo-munk** [reply]
You didn't hinder anything, that all were good improvements. I'll give the current dev sources another week to see if any problems come up, and if not I'll prepare the release version.

**Comment by sneakywumpus** [comment]
I made some changes to the z80pack homepage to fix internal link rot.
The attached file contains a patch and the new index.html produced by applying it to the current version.

[index.tar.gz](https://github.com/udo-munk/z80pack/files/9997488/index.tar.gz)

**Reply by udo-munk** [reply]
Thank you, just checked and uploaded to web site.

**Comment by udo-munk** [comment]
Are we ready for release or is something in the queue that should go into it?

**Reply by sneakywumpus** [reply]
I'm currently working on making the entire project follow a common coding style.
Especially the code by David McNaughton and Fred Weigel uses a different style.
I don't know what they will be saying if I mess with their code...

And the frontpanel library code is really unreadable. Is this still being maintained upstream?
The only thing I can find is an old site at https://sourceforge.net/projects/frontpanel which
has code from 2008 and some Windows patches in the Bugs section from 2015.

I will not touch civetweb, since this is an active upstream project.

I've settled on using astyle (https://astyle.sourceforge.net) for the formatting, since I can't get
clang-format to not mess up the all the nicely aligned end of line comments.

None of these tools is perfect, but they are very useful for an initial clean-up.

This is the .astylerc file I'm using:

```
style=kr
lineend=linux
indent=force-tab=8

keep-one-line-blocks
attach-closing-while
indent-preproc-define
min-conditional-indent=3
pad-comma
pad-header
unpad-brackets
squeeze-lines=1
align-pointer=name
attach-return-type
attach-return-type-decl
max-code-length=80

# don't use these for automatic formatting
# since they remove the space in "(int) 4"
# and change "(BYTE) *s" to "(BYTE) * s"
pad-oper
unpad-paren
```

Or should I stop this relatively massive and possibly disruptive undertaking?

**Comment by udo-munk** [comment]
So far I have avoided to mess to much with the formatting of others for some reasons.

Others might like the formatting they are using better and they migth want to support or at last read their code. More problematic is, if it is some library like frontpanel or the AMD MPU, these are used in multiple emulation projects. If owners of A come with one formatting, and then B come with another, what should the owner of the code do?

Probably you should contact them first and ask about this.

However, I'm fine with waiting for your clean ups, then this is through in the next release and verything neater.

**Comment by sneakywumpus** [comment]
OK, I'll regard iodevices/apu and frontpanel as upstream code, i.e. won't touch it style-wise.
Will also leave David's code as is for now, except when he added code to files started by you.

**Comment by udo-munk** [comment]
That's fine. I also reformatted code from Dave when it was started by me and he never complained. Just not sure about his stuff, don't really know what all is common in his esp32 machines. We can ask if he would like his code reformatted or not.

Do you think I should put something into Community Standards -> Contributing here, at least a hint that using astyle formatted code with your rules would be appreciated?

**Reply by sneakywumpus** [reply]
Would be nice if you asked.

I think it would be good to have a coding style standard for the project.

**Comment by udo-munk** [comment]
Excellent job Thomas! It seems that the support for bare metal machines is much improved. I can do real stuff now with the Raspberry Pico :-)

**Comment by udo-munk** [comment]
Now run the various machines with different CPU speed settings. The reported clock frequency for cpmsim is not correct because of the busy loop detection, which will put CPU to sleep while you are staring at some command prompt. All the others are pretty close now.

**Comment by sneakywumpus** [comment]
I'm thinking about removing CYGWIN support from the source. WSL seems to be all the rage today on Windows.
I'v just tested it on a Windows 10 VM and the dev branch z80pack runs fine in WSL, including the frontpanel graphics.

**Comment by udo-munk** [comment]
Fine with me, can go away. Since a while I'm using WSL only and it really is good enough nowadays, even on Windows 11 not broken.

**Reply by sneakywumpus** [reply]
I found a few more Cygwin hacks by going through the git history. What about the thread deadlock workarounds in imsai-vio and proctec-vdm? Are these necessary only for Cygwin? Can they go?

**Comment by udo-munk** [comment]
I first run into such problems with Cygwin, but there also were Linux systems, where the whole thread stuff got stuck. The problems came because of the difficulties to make X11 POSIX thread save. This also are not really hacks, there are just two ways to get all the threads stopped and the X11 window closed in the end. I don't know if this is fixed for good everywhere nowadays.

**Comment by udo-munk** [comment]
on OSX:

cc -O3 -std=c99 -D_DEFAULT_SOURCE  -Wall -Wextra -Wwrite-strings  -U_FORTIFY_SOURCE -DCONFDIR=\"/usr/local/share/altairsim/conf\" -DDISKSDIR=\"/usr/local/share/altairsim/disks\" -DBOOTROM=\"/usr/local/share/altairsim/roms\" -DFRONTPANEL -iquote . -I../../z80core -I../../iodevices -I../../frontpanel -I/opt/X11/include   -c -o simfun.o ../../z80core/simfun.c
../../z80core/simfun.c:112:2: warning: implicit declaration of function
      'clock_gettime' is invalid in C99 [-Wimplicit-function-declaration]
        clock_gettime(CLOCK_REALTIME, &ts);
        ^
../../z80core/simfun.c:112:16: error: use of undeclared identifier
      'CLOCK_REALTIME'
        clock_gettime(CLOCK_REALTIME, &ts);
                      ^
1 warning and 1 error generated.
make: *** [simfun.o] Error 1

**Reply by sneakywumpus** [reply]
It compiles fine on my MacBook running macOS Monterey 12.7.4 with Xcode 14.2. What version are you using?

**Comment by udo-munk** [comment]
Still 10.9.5, last one with floppy disk support, they removed it after that.

**Reply by sneakywumpus** [reply]
I just found out that clock_gettime is available since macOS 10.12 ...

**Reply by sneakywumpus** [reply]
Sorry meant 12.0

**Reply by sneakywumpus** [reply]
Forget it, my eyes get blurry... it is 10.12

**Reply by udo-munk** [reply]
Works again. I need a new one sometime, this is a 13" Pro from 2014.

**Comment by sneakywumpus** [comment]
Just a question: from where did you get the 2.1 version of the frontpanel code. I can't find it anywhere on the web.

**Comment by udo-munk** [comment]
This is because it is not available anywere else than in z80pack. This is the latest version I worked on with John, but he never updated his sourceforge repository anymore.

**Comment by udo-munk** [comment]
I had a look at my email archives, last time I talked to John was in 2017. No idea if he still can be reached, but I can try if you want. Then we can discuss with him how to proceed.

**Reply by sneakywumpus** [reply]
It just my compulsive :-) desire to reformat the code... it looks really wild to me.

**Comment by udo-munk** [comment]
I know, to me too. In the past I reformated code from others without asking, and the one or other complained. I also would complain if someone reformats my sources into something I don't like. So we should ask first.

**Comment by sneakywumpus** [comment]
Is https://github.com/sneakywumpus/z80pack/commit/39abc24df0b4591753ccd77f97389a15ee492d70 acceptable, or do want keeping every instruction i a separate function?

**Comment by udo-munk** [comment]
I want each instruction in it's own function. The reason is that you can anytime look what some instruction does without the need to dig through piles of macros and functions. I had the discussion already again and again in the past ;-)

**Reply by sneakywumpus** [reply]
Ok. No stress. :-)

**Comment by udo-munk** [comment]
Do you know Toledo's 8080 implementation? This one saves a great lot of memory and is fast, if you have some spare time look at it, is quite interesting.

**Reply by sneakywumpus** [reply]
Can't beat that. Crazy stuff.

**Reply by udo-munk** [reply]
Yes, he took that to the other extreme. Here with z80pack we take another approach. I know that is is used for teaching kids and young adults how a CPU works and how to program it in assembler. I even was asked to produce videos for kids to teach them that. I know that even kids seem to be able to understand, how this CPU fetches instructions, decodes them and does't it's stuff. This has to do with the software design decisions, because if you would give them Toledo's code, they would be, and stay, pretty clueless.

**Comment by udo-munk** [comment]
Forgot ...
The other reason is the usual space vs speed tradeoff. All CPU instructions are void, if common functions are used it will save space, but for the parameter passing it costs time.

**Reply by sneakywumpus** [reply]
That was one of the reasons I did this: the functions in simz80-ddcb.c are already taking a parameter (the displacement). I just changed the parameter to the complete address (IX + displacement).

**Reply by sneakywumpus** [reply]
Or should we change this, so that each function fetches the displacement itself?

**Reply by sneakywumpus** [reply]
Forget my last comment. The reason is for fetching it in op_ddcb_handle is obviously that the offset comes before the opcode in memory.

**Comment by udo-munk** [comment]
Yes, and also shows nicely what a Z80 is doing while decoding an instruction. In this case it finds that it needs to fetch another byte from memory and use it to calculate the target address.

**Comment by udo-munk** [comment]
Got new notebook
CPU executed 768249311 JP instructions in 3 seconds
clock frequency = 2560.83 MHz
Holy Cow!

**Reply by sneakywumpus** [reply]
Nice, my Intel Core i7-9700T PC running Arch Linux:
CPU executed 494668753 JP instructions in 3 seconds
clock frequency = 1648.90 MHz

**Comment by udo-munk** [comment]
With AMD Ryzen 7, Ubunto Linux under Windows 11/WSL:
CPU executed 496379907 JP instructions in 3 seconds
clock frequency = 1654.60 MHz
So almost the same.
This M3 from Apple really is fast, don't know what kind of Intel/AMD one would need to reach that.

**Comment by udo-munk** [comment]
If one interrupts a program running under control of the ICE the message we get just this:
Unknown error 9
Something wrong with the report stuff now.

**Comment by sneakywumpus** [comment]
Are the #include "../../iodevices/blah.h" and #include ".../../frontpanel/frontpanel.h" intentional? That is what the "-I" options in the Makefiles is for, so that one can just write #include "frontpanel.h".

**Comment by udo-munk** [comment]
No, this are relicts from using plain old UNIX System V makefiles, can go.

**Comment by sneakywumpus** [comment]
I'd like to rename memory.[hc] to memsim.[hc] (like iosim) to get rid of the conflict with <memory.h>  and -iquote .

**Reply by sneakywumpus** [reply]
Or perhaps simio.[hc] & simmem.[hc] like simctl.c

**Comment by udo-munk** [comment]
Fine with me, this are just names I used when it was implemented, more logical names will help others.

**Comment by sneakywumpus** [comment]
In my quest to make z80pack installable I'm intending to make FRONTPANEL runtime opt-out and HAS_NETSERVER runtime opt-in, i.e. an option '-F' to disable the front panel and '-p <port>' to disable the front panel and start a net server at port <port>.

Leaving out FRONTPANEL and HAS_NETSERVER from sim.h would still completely exclude the corresponding code.

Opinions?

**Reply by udo-munk** [reply]
Seems to be OK, this would only change the commands to produce one setup or another, and we write this into the documentation then.

**Reply by sneakywumpus** [reply]
Should the HAS_DISKMANAGER code path only be used when the web server is enabled?

**Reply by sneakywumpus** [reply]
I.e. the disk manager is only used when the simulator is run with "-n"?

**Reply by udo-munk** [reply]
The web server won't see disk drives, if diskmannager is not included, this was necessary to change disks in the web frontend. Without usage is optional, machine can be used with or without. What works is if disk manger always is used, so I think this is OK, there is no good reason to not use it.

**Reply by sneakywumpus** [reply]
In imsai-fif.c the large drive I is only supported without HAS_DISKMANAGER

**Reply by udo-munk** [reply]
OK, then there was a reason :-(
The problem is, floppy disks can be changed anytime by the user, a hard disk of course not, this is a fixed device. Dave has provisions for this in his ESP32 machines, but I don't know what.  We need to make the hard disk somehow available without going through diskmannager.

**Reply by udo-munk** [reply]
Same problem likely with the Z1, even has 3 hard disks.

**Reply by udo-munk** [reply]
No problems with the Z1, has separate FD and HD controllers, each handling their drives.

**Comment by udo-munk** [comment]
Looked through the code for IMSAI, can easily be fixed. Harddisk needs separate handling is not a floppy disk, so needs to go out of the disks[] array. Then all this #ifdef LARGEDISK can go, no one wants a system without a hard disk. Also all this #ifdef HAS_DISKMANAGER can go, we always use it.

**Comment by udo-munk** [comment]
Just testing, after the latest changes I cannot access disks/drivei.dsk on the IMSAI anymore at all:

A>dir i:
Bdos Err On I: Bad Sector

**Comment by udo-munk** [comment]
Forgot to make newest bits, still was running the old code with HD disabled because diskmanager, looks fine now.

**Comment by udo-munk** [comment]
Copyright and history entry in imsai-fif.[ch] need update.

**Comment by sneakywumpus** [comment]
The 8080/Z80-core source code is starting to look messy (yours truly ducks into a corner...), and I'm inclined to, as you suggested, implemented a second selectable set of cores more optimized for size/speed and clean-up the original ones with only the UNDOC_*/FAST_BLOCK options remaining.

**Comment by udo-munk** [comment]
No problem, you see what I meant. Completely separated cores would be better, then one can have a readable and easy to understand/debuggable one, while others can be more advanced or optimized for some special purpose.

**Reply by sneakywumpus** [reply]
Or should we even remove the UNDOC_FLAGS and UNDOC_IALL options from the original core to leave this as a by the book implementation?

**Reply by sneakywumpus** [reply]
As it was, before I came along...

**Comment by udo-munk** [comment]
Implementing the undocumented flags is completely useless. All one can do is PUSH AF and then look at the bits, of course there are no instructions that can branch on them, or anything. The only effect it has is a huge pile of useless code consuming memory and time, to make one instruction exerciser happy. Way back then I also had the discussion with Frank Cringle, who did understand my point and agreed to make the test of the undocumented flag bits an option and not a must. Not needed, not useful, can go if you ask me. Maybe wait a while to hear other opinions.

**Reply by sneakywumpus** [reply]
I was wondering a bit why you did accept my push request with this stuff, we already had this discussions before... But now that the code had passed all available tests and I had my "Erfolgserlebnis"  it doesn't really have any practical use. I'll remove it. What about UNDOC_IALL instructions?

**Reply by udo-munk** [reply]
Because now you had your "Erfolgserlebnis" and know exactly why I implemented it as it was.

**Comment by udo-munk** [comment]
The undocumented instructions I had implemented were those really used by some old software, TDL mostly did this. I added one after another until there was no software anymore that would trap. So same here, instructions that never ever are used by anything waste memory and worse, time to keep them working correct.

**Reply by sneakywumpus** [reply]
Oops didn't read this before replying. Will also remove.

**Comment by udo-munk** [comment]
In cpmsim the measured runtime cannot be correct:

System halted
CPU ran 71 ms and executed 10862565 t-states
Clock frequency 151.73 MHz

I typed a few commands, and sure needed several seconds for that, I am not this fast ;-)

**Comment by udo-munk** [comment]
Ah, is corrected for the time the busy wait sleeped the CPU.

**Comment by udo-munk** [comment]
On the IMSAI:

System halted
CPU ran 4681 ms and executed 33731009 t-states
Clock frequency 7.20 MHz

This is not possible. Happens if one resets the CPU from front panel. Probably because reset_cpu() sets R to 0.
